File: 2015/day3/part1.cpp
#include <iostream>
#include <vector>

int main(int argc, char *argv[])
{
	std::string line;
	std::getline(std::cin, line);
	int x_count = 1, y_count = 1;
	for (const char &c : line) {
		if (c == '<' || c == '>') x_count++;
		else if (c == '^' || c == 'v') y_count++;
	}
	std::vector<std::vector<int> > grid;
	for (unsigned i = 0; i < x_count; i++) {
		std::vector<int> row(y_count, 0);
		grid.push_back(row);
	}
	int current_x = (grid[0].size() % 2 == 0 ? grid[0].size() / 2 + 1 : grid[0].size() / 2);
	int current_y = (grid.size() % 2 == 0 ? grid.size() / 2 + 1 : grid.size() / 2);
	grid[current_y][current_x] = 1;
	for (const char &c : line) {
		if (c == '^') {
			current_y++;
			grid[current_y][current_x]++;
		}
		else if (c == 'v') {
			current_y--;
			grid[current_y][current_x]++;
		}
		else if (c == '>') {
			current_x++;
			grid[current_y][current_x]++;
		}
		else if (c == '<') {
			current_x--;
			grid[current_y][current_x]++;
		}
	}
	int count = 0;
	for (unsigned i = 0; i < grid.size(); i++) {
		for (unsigned j = 0; j < grid[0].size(); j++) {
			if (grid[i][j] > 0) {
				count++;
			}
		}
	}
	std::cout << count << '\n';
	return 0;
}

File: 2015/day3/part1.swift
func part1(_ input: String) -> Int {
    var xCount = 1, yCount = 1
    for dir in input {
        switch dir {
            case "<":
            fallthrough
            case ">":
            xCount += 1
            case "^":
            fallthrough
            case "v":
            yCount += 1
            default: break
        }
    }
    var grid = (0..<xCount).map { _ in return Array<Int>(repeating: 0, count: yCount) }
    var currX = (grid[0].count % 2 == 0 ? grid[0].count / 2 + 1 : grid[0].count / 2)
    var currY = (grid.count % 2 == 0 ? grid.count / 2 + 1 : grid.count / 2)
    grid[currY][currX] = 1
    for dir in input {
        switch dir {
            case "^":
            currY += 1
            case "v":
            currY -= 1
            case "<":
            currX += 1
            case ">":
            currX -= 1
            default: break
        }
        grid[currY][currX] += 1
    }
    let ret = grid.reduce(0) { (acc: Int, row: [Int]) -> Int in
        return acc + row.reduce(0) { (innerAcc: Int, house: Int) -> Int in
            return innerAcc + (house > 0 ? 1 : 0)
        }
    }
    return ret
}